src/Random/Random.dfy

```diff
@@ -0,0 +1,22 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT
+ *******************************************************************************/
+
+module {:options "-functionSyntax:4"} Dafny.Random {
+
+    // Return an arbitrary boolean value.
+    predicate {:extern "DafnyLibraries.Random", "nextBool"} nextBool()
+
+    // Return an arbitrary integer value in the range [0, bound).  If the optional
+    // bound is omitted, return an arbitrary nonnegative integer (generally limited
+    // to a 64-bit integer value when compiled).
+    function {:extern "DafnyLibraries.Random", "nextInt"} nextInt(bound: int := 0): (value: int)
+        ensures 0 <= value
+        ensures bound > 0 ==> value < bound
+
+    // Return an arbitrary real value in the range [0,1).
+     function {:extern "DafnyLibraries.Random", "nextReal"} nextReal(): (value: real)
+        ensures 0.0 <= value < 1.0
```

Comment:
I don't think such a function makes sense because a random real number is irrational with probability 1, but `real` numbers in Dafny are translated to rational numbers in the target language. It is not even clear what a "uniform distribution" on the rational numbers would be.

So my suggestion: What we could have instead is a uniform distribution on `{0, 1/(2^k), 2/(2^k), ..., 2^k/(2^k)}` for a given `k` as an argument. This is probably reasonably close to what other languages do when they give a you a uniform floating-point value between 0 and 1.

Reply:
You're right in that it satisfies the post-condition. But I would argue that the name `Random.nextReal` creates an expectation that is not captured by the post-condition. Returning 0.5 deterministically would also satisfy the post-condition, but would probably be unexpected for users.